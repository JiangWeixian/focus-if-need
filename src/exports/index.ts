/**
 * @fileoverview focus if need
 * - automatic focus element if avaliable
 * - TODO: if active element changed, should automatic focus on last focused element
 */

import { createFocusIfNeed } from '../create'

export { createFocusIfNeed } from '../create'
export type { FocusIfNeedHookMap } from '../create'
export const focusIfNeed = createFocusIfNeed()
