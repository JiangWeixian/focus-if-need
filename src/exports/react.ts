import { useEffect, useRef } from 'react'

import { focusIfNeed } from './index'

export const useFocus = <T = any>(id: string) => {
  const ref = useRef<T>(null)
  useEffect(() => {
    const { clear } = focusIfNeed.focus(id, ref)
    return () => {
      clear()
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [id])
  return {
    ref,
  }
}

export { focusIfNeed } from './index'
