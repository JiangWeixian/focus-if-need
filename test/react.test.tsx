import TestRenderer from 'react-test-renderer'
import {
  afterEach,
  beforeEach,
  describe,
  expect,
  test,
} from 'vitest'

import { useFocus } from '@/exports/react'

import type { ReactTestRenderer, ReactTestRendererJSON } from 'react-test-renderer'

function toJson(component: ReactTestRenderer) {
  const result = component.toJSON()
  expect(result).toBeDefined()
  expect(result).not.toBeInstanceOf(Array)
  return result as ReactTestRendererJSON
}

describe('react', () => {
  let node: HTMLDivElement
  beforeEach(() => {
    node = document.createElement('div')
    document.body.appendChild(node)
  })

  afterEach(() => {
    document.body.removeChild(node)
    node = null!
  })

  test('useFocus should work', async () => {
    let focused = false
    const Page = () => {
      const { ref } = useFocus<HTMLInputElement>('main')
      console.log(ref)
      return (
        <div>
          <input
            ref={ref}
            placeholder="placeholder"
          />
        </div>
      )
    }

    let renderer: ReactTestRenderer
    await TestRenderer.act(() => {
      renderer = TestRenderer.create(
        <Page />,
        {
          createNodeMock: (element) => {
            if (element.type === 'input') {
              // mock a focus function
              return {
                focus: () => {
                  focused = true
                },
              }
            }
            return null
          },
        },
      )
    })

    const tree = toJson(renderer!)
    expect(tree).toMatchSnapshot()
    expect(focused).toBe(true)
  })
})
