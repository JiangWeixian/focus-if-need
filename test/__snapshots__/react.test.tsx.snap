// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`react > useFocus should work 1`] = `
<div>
  <input
    onFocus={[Function]}
    placeholder="placeholder"
  />
</div>
`;
